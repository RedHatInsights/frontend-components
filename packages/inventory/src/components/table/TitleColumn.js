/* eslint-disable camelcase */
/* eslint-disable react/prop-types */
import React from 'react';

/**
 * Helper function to proprly calculate what to do when user clicks on first cell.
 * Either full redirect if used with ctrl button or `onRowClick` from props is used.
 * @param {*} event html event, to find out if meta key was clicked.
 * @param {*} key inventory UUID.
 * @param {*} props additional props from `EntityTable` - loaded, onRowClick and noDetail.
 */
const onRowClick = (event, key, { loaded, onRowClick: rowClick, noDetail }) => {
    if (loaded && !noDetail) {
        const isMetaKey = (event.ctrlKey || event.metaKey || event.which === 2);
        if (isMetaKey) {
            const url = new URL(`./${key}`, location.href);
            window.open(url.href);
        } else if (rowClick) {
            rowClick(event, key, isMetaKey);
        }
    }
};

/**
 * Helper function to generate first cell in plain inventory either with clickable detail or just data from attribut.
 * This is later on used in redux in `renderFunc`.
 * @param {React.node} data React node with information that will be shown to user as column title.
 * @param {string} id inventory UUID, used to navigate to correct URL.
 * @param {*} item row data, holds every information from redux store for currecnt row.
 * @param {*} props additional props passed from `EntityTable` - holds any props passed to inventory table.
 */
const TitleColumn = (data, id, item, props) => (
    <div className="ins-composed-col">
        <div>{item?.os_release}</div>
        <div className={props?.noDetail ? 'ins-m-nodetail' : ''}>
            { props?.noDetail ?
                data :
                <a
                    widget="col"
                    href={ `${location.pathname}${location.pathname.substr(-1) === '/' ? '' : '/'}${id}` }
                    onClick={ event => {
                        event.preventDefault();
                        event.stopPropagation();
                        onRowClick(event, id, props);
                    }}
                >
                    { data }
                </a> }
        </div>
    </div>
);

export default TitleColumn;
