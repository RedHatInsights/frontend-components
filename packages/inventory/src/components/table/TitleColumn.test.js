/* eslint-disable camelcase */
import React from 'react';
import TitleColumn from './TitleColumn';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('TitleColumn', () => {
    it('should render correctly with NO data', () => {
        const Cmp = () => TitleColumn();
        const wrapper = mount(<Cmp />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render correctly with data', () => {
        const Cmp = () => TitleColumn('something', 'some-id', { os_release: 'os_release' });
        const wrapper = mount(<Cmp />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render correctly no detail with data', () => {
        const Cmp = () => TitleColumn('something', 'some-id', { os_release: 'os_release' }, { noDetail: true });
        const wrapper = mount(<Cmp />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    describe('API', () => {
        const open = jest.fn();

        beforeEach(() => {
            Object.defineProperty(window, 'open', {
                writable: true,
                value: open
            });
        });

        it('should call onClick', () => {
            const onClick = jest.fn();
            const Cmp = () => TitleColumn(
                'something',
                'some-id',
                { os_release: 'os_release' },
                { onRowClick: onClick, loaded: true }
            );
            const wrapper = mount(<Cmp />);
            wrapper.find('a').first().simulate('click');
            expect(onClick).toHaveBeenCalled();
        });

        it('should NOT call onClick', () => {
            const onClick = jest.fn();
            const Cmp = () => TitleColumn(
                'something',
                'some-id',
                { os_release: 'os_release' },
                { onRowClick: onClick }
            );
            const wrapper = mount(<Cmp />);
            wrapper.find('a').first().simulate('click');
            expect(onClick).not.toHaveBeenCalled();
        });

        it('should NOT call onClick and call global open', () => {
            const onClick = jest.fn();
            const Cmp = () => TitleColumn(
                'something',
                'some-id',
                { os_release: 'os_release' },
                { onRowClick: onClick, loaded: true }
            );
            const wrapper = mount(<Cmp />);
            wrapper.find('a').first().simulate('click', { ctrlKey: true });
            expect(onClick).not.toHaveBeenCalled();
            expect(open).toHaveBeenCalled();
        });
    });
});
