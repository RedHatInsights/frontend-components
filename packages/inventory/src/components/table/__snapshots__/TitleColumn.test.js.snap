// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TitleColumn should render correctly no detail with data 1`] = `
<Cmp>
  <div
    className="ins-composed-col"
  >
    <div>
      os_release
    </div>
    <div
      className="ins-m-nodetail"
    >
      something
    </div>
  </div>
</Cmp>
`;

exports[`TitleColumn should render correctly with NO data 1`] = `
<Cmp>
  <div
    className="ins-composed-col"
  >
    <div />
    <div
      className=""
    >
      <a
        href="/undefined"
        onClick={[Function]}
        widget="col"
      />
    </div>
  </div>
</Cmp>
`;

exports[`TitleColumn should render correctly with data 1`] = `
<Cmp>
  <div
    className="ins-composed-col"
  >
    <div>
      os_release
    </div>
    <div
      className=""
    >
      <a
        href="/some-id"
        onClick={[Function]}
        widget="col"
      >
        something
      </a>
    </div>
  </div>
</Cmp>
`;
